### PR TITLE
Support JsonExtensionData in STJ source-gen

### LINKED ISSUE #58273: Source generator doesn't support [JsonExtensionData]
The `[JsonExtensionData]` attribute is not supported and silently fails.

Expect that the parser detects that attribute and passes that information\metadata to the serializer (best long-term), or we generate a design-time error (to help scope V6).

Repro:
```cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Test;

POCO? poco = JsonSerializer.Deserialize("{\"Hello\":42}", MyContext.Default.POCO);
// Non-source gen works:
// POCO? poco = JsonSerializer.Deserialize<POCO>("{\"Hello\":42}");

// Throws since Overflow is null
Console.WriteLine(poco.Overflow["Hello"]);

namespace Test
{

    [JsonSerializable(typeof(POCO))]
    public partial class MyContext : JsonSerializerContext
    {
    }

    public partial class POCO
    {
        [JsonExtensionData]
        public Dictionary<string, JsonElement> Overflow { get; set; }
    }
}
```

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/58273. We should port this to 6.0.

### CODE REVIEW

File: src/libraries/System.Text.Json/gen/Reflection/TypeExtensions.cs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Linq;
+using System.Text.Json.SourceGeneration;
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/gen/Reflection/TypeExtensions.cs:7]: Is this change necessary?

File: src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs [hunk 2]
```diff
@@ -221,6 +229,11 @@ private bool FastPathIsSupported()
         {
             if (ClassType == ClassType.Object)
             {
+                if (ExtensionDataPropertyTypeSpec != null)
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs:232]: Why can't the FastPath support this?

Comment (layomia) [reply src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs:232]: Extension data properties (dictionaries) can have `typeof(object)` values which is a polymorphic scenario currently not supported in fast-path. Also, the ext. properties can be `JsonObject` - I spent some time prototyping support for this but the PR started getting too complex and large.

We always had `JsonExtensionData` as "not supported in fast path for 6.0" - https://github.com/dotnet/runtime/issues/51945#issuecomment-827795787. Would have be nice to get it in now, but I don't think it's worth the extra complexity at this stage.

File: src/libraries/System.Text.Json/gen/Reflection/TypeExtensions.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/gen/Reflection/TypeExtensions.cs:7]: Likely not, will fix.

File: src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs:232]: Do you think it makes sense to log this as an issue, explaining the limitations and concerns? At least then it will be written down somewhere.

File: src/libraries/System.Text.Json/gen/Resources/Strings.resx [hunk 3]
```diff
@@ -141,4 +141,16 @@
...
     <value>Type has multiple constructors annotated with JsonConstructorAttribute.</value>
   </data>
+  <data name="MultipleJsonExtensionDataAttributeFormat" xml:space="preserve">
+    <value>Type '{0}' has multiple members annotated with 'JsonExtensionDataAttribute'.</value>
+  </data>
+  <data name="MultipleJsonExtensionDataAttributeTitle" xml:space="preserve">
+    <value>Type has multiple members annotated with JsonExtensionDataAttribute.</value>
+  </data>
+  <data name="DataExtensionPropertyInvalidFormat" xml:space="preserve">
+    <value>The data extension property '{0}.{1}' does not match the required signature of 'IDictionary&lt;string, JsonElement&gt;', 'IDictionary&lt;string, object&gt;' or 'JsonObject'.</value>
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/gen/Resources/Strings.resx:151]: It has to be `IDictionary`? It can't be just `Dictionary`?

File: src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs [hunk 4]
```diff
@@ -58,6 +58,10 @@ internal class TypeGenerationSpec
 
         public TypeGenerationSpec? NullableUnderlyingTypeMetadata { get; private set; }
 
+        public string? RuntimeTypeRef { get; private set; }
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs:61]: It might make sense to stick a comment on this property explaining what it is and what it is used for. At first read, I don't really understand what a `RuntimeTypeRef` is.

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs [hunk 5]
```diff
@@ -963,6 +963,7 @@ public static partial class JsonMetadataServices
         public static System.Text.Json.Serialization.JsonConverter<int> Int32Converter { get { throw null; } }
         public static System.Text.Json.Serialization.JsonConverter<long> Int64Converter { get { throw null; } }
         public static System.Text.Json.Serialization.JsonConverter<System.Text.Json.JsonElement> JsonElementConverter { get { throw null; } }
+        public static System.Text.Json.Serialization.JsonConverter<System.Text.Json.Nodes.JsonObject> JsonObjectConverter { get { throw null; } }
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:966]: Is this supposed to be part of this change? It seems unrelated...

File: src/libraries/System.Text.Json/tests/Common/ExtensionDataTests.cs [hunk 6]
```diff
@@ -0,0 +1,1475 @@
...
+        {
+            [JsonExtensionData]
+            public Dictionary<string, object> MyOverflow { get; set; }
+
+            public Dictionary<string, object> ActualDictionary { get; set; }
+        }
+
+        [Fact]
+#if BUILDING_SOURCE_GENERATOR_TESTS
+        [ActiveIssue("Compile-time errors thrown for these scenarios.")]
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/tests/Common/ExtensionDataTests.cs:1149]: Should we be logging issues for these disabled tests?

File: src/libraries/System.Text.Json/gen/Resources/Strings.resx (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Text.Json/gen/Resources/Strings.resx:151]: It has to implement `IDictionary<,>` or be `JsonObject`. Can update the message. It's the same one we have in the reflection serializer.

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs (hunk 5 above)
Comment (layomia) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:966]: Per https://github.com/dotnet/runtime/pull/58912#discussion_r705797719, `JsonObject` can be used as ext. data. We need this to support that scenario (also for upcoming support for S.T.Json.Nodes types - https://github.com/dotnet/runtime/issues/45448#issuecomment-915449217).

Comment (eerhardt) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:966]: I must be missing something, but I'm not seeing where the generated code calls this. Can you point me to where it is used?

File: src/libraries/System.Text.Json/gen/Resources/Strings.resx (hunk 3 above)
Comment (eerhardt) [reply src/libraries/System.Text.Json/gen/Resources/Strings.resx:151]: Ah, that makes more sense. The way it is currently written, it sounds like the property's type has to strictly be `IDictionary`.

File: src/libraries/System.Text.Json/gen/Resources/Strings.resx [hunk 7]
```diff
@@ -141,4 +141,16 @@
   <data name="MultipleJsonConstructorAttributeTitle" xml:space="preserve">
     <value>Type has multiple constructors annotated with JsonConstructorAttribute.</value>
   </data>
+  <data name="MultipleJsonExtensionDataAttributeFormat" xml:space="preserve">
+    <value>Type '{0}' has multiple members annotated with 'JsonExtensionDataAttribute'.</value>
```
Comment (danmoseley) [inline src/libraries/System.Text.Json/gen/Resources/Strings.resx:145]: It's not consistent whether you quote time names or not. Two messages here have quotes and two do not. I'd say in general we don't quote unless possibly if it's an insertion.
```suggestion
    <value>Type '{0}' has multiple members annotated with JsonExtensionDataAttribute.</value>
```

File: src/libraries/System.Text.Json/gen/Resources/Strings.resx (hunk 3 above)
Comment (danmoseley) [reply src/libraries/System.Text.Json/gen/Resources/Strings.resx:151]:
```suggestion
    <value>The data extension property '{0}.{1}' does not match the required signature of IDictionary&lt;string, JsonElement&gt;, IDictionary&lt;string, object&gt; or JsonObject.</value>
```

File: src/libraries/System.Text.Json/gen/Resources/Strings.resx (hunk 7 above)
Comment (danmoseley) [reply src/libraries/System.Text.Json/gen/Resources/Strings.resx:145]: There's some others quoted in this file as well. Do what you will :)

File: src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs:232]: Yup I opened https://github.com/dotnet/runtime/issues/58943.

File: src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs (hunk 4 above)
Comment (layomia) [reply src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs:61]: I added a comment:

```
/// <summary>
/// Supports deserialization of extension data dictionaries typed as I[ReadOnly]Dictionary<string, object/JsonElement>.
/// Specifies a concrete type to instanciate, which would be Dictionary<string, object/JsonElement>.
/// </summary>
```

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs (hunk 5 above)
Comment (layomia) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:966]: The code that calls it is generated implicitly on this line: https://github.com/dotnet/runtime/blob/f4afd921e20fc7e2d7f17543784e8d453f5bf57e/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs#L292

`{typeFriendlyName}Converter` resolves to `JsonObjectConverter`.

File: src/libraries/System.Text.Json/tests/Common/ExtensionDataTests.cs (hunk 6 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Common/ExtensionDataTests.cs:1149]: Created https://github.com/dotnet/runtime/issues/58945. Will link it in source.

File: src/libraries/System.Text.Json/gen/Resources/Strings.resx (hunk 7 above)
Comment (layomia) [reply src/libraries/System.Text.Json/gen/Resources/Strings.resx:145]: I thought that we also quote names if we don't want translation to convert them, e.g. type names.

File: src/libraries/System.Text.Json/gen/Resources/Strings.resx (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Text.Json/gen/Resources/Strings.resx:151]: I've updated the exception message. I had a question on quotes in https://github.com/dotnet/runtime/pull/58912#discussion_r706239039.

File: src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs [hunk 8]
```diff
@@ -58,6 +58,14 @@ internal class TypeGenerationSpec
 
         public TypeGenerationSpec? NullableUnderlyingTypeMetadata { get; private set; }
 
+        /// <summary>
+        /// Supports deserialization of extension data dictionaries typed as I[ReadOnly]Dictionary<string, object/JsonElement>.
+        /// Specifies a concrete type to instanciate, which would be Dictionary<string, object/JsonElement>.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs:63]:
```suggestion
        /// Specifies a concrete type to instantiate, which would be Dictionary<string, object/JsonElement>.
```

File: src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs (hunk 4 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs:65]: I'm assuming this would not be used if the extension data property is of type `JsonObject`, right?

File: src/libraries/System.Text.Json/tests/Common/ConstructorTests/ConstructorTests.Cache.cs [hunk 9]
```diff
@@ -11,9 +11,6 @@ public abstract partial class ConstructorTests
     {
         [Fact]
         [OuterLoop]
-#if BUILDING_SOURCE_GENERATOR_TESTS
-        [ActiveIssue("Needs JsonExtensionData support.")]
-#endif
         public async Task MultipleThreadsLooping()
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/ConstructorTests/ConstructorTests.Cache.cs:14]: Nit: what is this test meant to be testing? If it's ensuring that the serialization infrastructure can cope with many serialization in parallel, that is generally handled by xunit's capacity to run multiple tests in parallel. If it's testing something else, it should be refelected in the test name.

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/System.Text.Json/gen/Resources/Strings.resx (hunk 7 above)
Comment (danmoseley) [reply src/libraries/System.Text.Json/gen/Resources/Strings.resx:145]: In these cases there is no way anyone would attempt to translate them as they're obviously not words. Maybe if it was `Date` but then there is a way to annotate the resource so they don't localize a part...

**Comment by eiriktsarpalis** [comment]
@layomia looks like we need to get #58983 merged before we can start backporting this one. I expect the same will likely apply to #58993.

cc @danmoseley
